### ISSUE TITLE
nodejs server OpenAPI 3.0 links

### ISSUE THREAD

**Issue by gusmonod** [Mermade/openapi-codegen#19]
First of all, thanks a lot for making a code generator compatible for OpenAPI 3.0.x and nodejs-server.

I noticed that (at least in version 1.3.4) there are a couple features of OpenAPI that do not seem to be fully working, and I couldn't find any reference to this in your documentation.

Are you interested in implementing these features? Did you know that they were part of the OpenAPI 3.0.0 specification?

1. [Response links](https://swagger.io/specification/#linkObject)

OpenAPI 3.0.x allows for an additional 'links' property in the response, that is described in the above link. The failure I had was for the following (brief but complete) example:

```yaml
openapi: 3.0.0
info:
  title: 'title'
  description: 'description'
  version: 0.0.1


servers:
- url: 'http://localhost:5000/api/v0'
  description: 'Local dev server'


paths:
  /pets:
    get:
      operationId: 'getAllPets'
      summary: 'Get all pets'
      description: 'Retrieve the latest pets'
      responses:
        '200':
          description: 'Successful: list of pets'
          content:
            application/json:
              schema:
                type: 'object'
                properties:
                  pets:
                    type: 'array'
                    items:
                      type: 'object'
                      properties:
                        petId:
                          type: 'string'
          links:
            petId:
              operationId: 'getPetById'
              parameters:
                petId: '$response.body#/pets/petId'

  /pets/{petId}:
    get:
      operationId: 'getPetById'
      summary: 'Get a pet by its ID'
      parameters:
        - in: 'path'
          name: 'petId'
          schema:
            type: 'string'
          required: true
          description: 'ID of the pet to get'
      responses:
        '200':
          description: 'Success'
          content:
            application/json:
              schema:
                type: 'object'
                properties:
                  pet:
                    $ref: '#/components/schemas/Pet'

components:
  schemas:
    Pet:
      type: 'object'
      properties:
        name:
          type: 'string'
```

2. [Response Patterned Fields](https://swagger.io/specification/#httpCodes)

Another (arguably simpler) feature that OpenAPI 3.0.0 introduced is the "Patterned Fields". They are used in order to provide a "catch all" nest for HTTP status codes.

For example, one could write:

```yaml
paths:
  /pets:
    get:
      operationId: 'getAllPets'
      responses:
        '200':
          description: 'Success'
        '4XX':
          description: 'Other failure than 404'
        '404':
           description: 'Not found'
```

**Comment by MikeRalphson** [comment]
Thanks for the kind words!

>I noticed that (at least in version 1.3.4) there are a couple features of OpenAPI that do not seem to be fully working, and I couldn't find any reference to this in your documentation. Are you interested in implementing these features? 

Because we take the templates unmodified from `swagger-codegen`, and those templates are designed originally to work with OpenAPI / Swagger 2.0, almost by definition, new features from OpenAPI 3.0.x like `links` and `callbacks` are hard to support.

Where it is possible to find a template model representation of a 3.0.x feature which allows its use, I'm definitely all for that.

Patterned response keys should be passed through like any other status code at the moment, but I can see perhaps the need to have a switch to toggle that behaviour?

> The failure I had [with links]

This is where it gets interesting, can you detail what failed and what error messages you received? If so we can look at improving things.

> Did you know that they were part of the OpenAPI 3.0.0 specification?

Yep, being on the OpenAPI TSC kind of helps. :slightly_smiling_face:

**Comment by gusmonod** [comment]
> Because we take the templates unmodified from `swagger-codegen`, and those templates are designed originally to work with OpenAPI / Swagger 2.0, almost by definition, new features from OpenAPI 3.0.x like `links` and `callbacks` are hard to support. Obviously if you are using your own / modified templates, that's slightly different.

Ok, I now understand the scope of this project a bit better. I guess it was wishful thinking, but I originally thought that this project was intending to support **all** of the OpenAPI 3.0.x specification.

I can't believe I forgot to add the error messages! Anyway, here they are:

```
$ node ~/openapi-codegen-1.3.4/cg.js --verbose nodejs oas.yml && (cd out/nodejs && npm start)
Loaded config nodejs    
Loaded definition oas.yml
{ level: 'Valid',                                
  elementType: 'Context',
  elementId: 'None',
  message: 'No validation errors detected' }          
Processing index.js
[...]
npm installation
[...]
> node index.js

API Errors:

  #/paths/~1pets/get/responses/200: Not a valid response definition
    #/paths/~1pets/get/responses/200: Additional properties not allowed: links
    #/paths/~1pets/get/responses/200: Missing required property: $ref

1 error and 0 warnings
[...]
npm ERR! Exit status 1
```

I have removed a couple section (indicated by [...]) for brevity, but if you believe they may be interesting, I can definitely add them. Also, I will update my original comment so that the entire content of the `oas.yml` file is visible to you.

As you can see, the validation and generation process both work fine, but at the moment of running `index.js`, those two errors arise.

**Comment by MikeRalphson** [comment]
> I guess it was wishful thinking, but I originally thought that this project was intending to support all of the OpenAPI 3.0.x specification.

With custom/modified templates, the whole of the OAS 3 definition is available for use, but especially in the case of the `nodejs` templates, which are a very thin wrapper on top of  `swagger-tools` which is actually doing all the work, if `swagger-tools` doesn't support links (because it's OAS 2.0 only) then there is little we can do.

I've made a fix to the OAS 3.0.x to 2.0 downconverter which is used to pass a `swagger.yaml` definition into `swagger-tools`, this simply renames the response `links` so they don't cause a validation failure.

If you can clone and use the project from the github repository, this fix is there now. There are a couple of outstanding issues holding up me cutting a new npmjs release, so I'll leave this issue open until then.

**Comment by MikeRalphson** [comment]
Version 1.4.1 released with this fix. Thanks again.
